In [ ]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

%matplotlib inline

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier1
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix
from sklearn.model_selection import cross_validate
import multiprocessing
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score
from sklearn.svm import LinearSVC
import warnings
warnings.filterwarnings('ignore')

### Answer a)

In [ ]:
with open('Data_Q1.txt','r') as file:
    lines = file.readlines()

no_of_samples=len(lines)

y=np.zeros(no_of_samples)

d={}
m=[]

for i in range(no_of_samples):
  l=lines[i].split()
  y[i]=l[0]
  X_dict={}
  for j in l[1:]:
    k=j.split(':')
    X_dict[int(k[0])]=int(k[1])
  m.append(max(X_dict.keys()))
  d[i+1]=X_dict

no_of_features=max(m)

X=np.zeros((no_of_samples, no_of_features))

for k2,v in d.items():
  for k1,v1 in v.items():
    X[k2-1][k1-1]=v1

print(X)
print(y)

[[1. 1. 1. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 1. 1. 1.]
 [0. 0. 0. ... 0. 0. 0.]]
[ 1. -1. -1. ...  1. -1.  1.]


In [ ]:
print((no_of_samples, no_of_features))

(4143, 54877)


In [ ]:
print(X.shape)

(4143, 54877)


In [ ]:
my_matrix=X

In [ ]:
from sklearn.decomposition import PCA

my_model = PCA(n_components=520)
X_new=my_model.fit_transform(my_matrix)

print (my_model.explained_variance_ratio_.cumsum())

[0.06151249 0.10554605 0.1291997  0.14915661 0.16583007 0.17891754
 0.19130718 0.20264855 0.2128589  0.22274433 0.23229439 0.24062825
 0.2482886  0.25581608 0.26315092 0.26967828 0.27586312 0.28186658
 0.2876791  0.29331067 0.29892204 0.30427433 0.30956145 0.31472194
 0.31966877 0.32443239 0.3289975  0.33352215 0.33797888 0.34224001
 0.34643139 0.35059192 0.3546258  0.35854086 0.36240983 0.36623747
 0.3699218  0.37357193 0.37711375 0.38056736 0.38399015 0.38733268
 0.39064541 0.39388086 0.39708878 0.40021163 0.40328667 0.40633915
 0.40936083 0.41232849 0.41528644 0.41815654 0.42097173 0.42375116
 0.42643715 0.42908486 0.43169565 0.43423838 0.43675617 0.43927008
 0.44175135 0.44418376 0.44661012 0.44897991 0.45132254 0.45365853
 0.45598168 0.45822013 0.46044395 0.46263343 0.46480956 0.46695911
 0.46907389 0.47116247 0.47323623 0.47526996 0.47728671 0.4792883
 0.48127499 0.48324654 0.48520248 0.48714164 0.4890487  0.49094193
 0.49282388 0.49465412 0.49646814 0.4982701  0.50006522 0.50184

In [ ]:
print(X_new)

[[-3.16355559e+00  1.35449233e+00  1.22609270e-01 ...  1.73978686e-02
   2.22212971e-02 -6.84901145e-03]
 [ 2.94461102e+00 -1.85303372e+00 -3.42356930e+00 ...  1.73635200e-01
   7.04242106e-02  2.76140411e-02]
 [-2.42472556e+00  1.07460981e+00 -9.15473283e-02 ... -5.85720272e-02
   6.17565303e-02  2.56509028e-01]
 ...
 [ 8.89478955e-01 -9.01727203e-01  2.04859233e+00 ...  3.24235490e-01
  -2.87580279e-01 -3.15313683e-01]
 [-1.75517906e+00  5.72070548e-01  1.17252505e-01 ...  1.67895041e-01
   5.39460948e-02 -1.55764081e-01]
 [ 1.22167361e+01  1.40410531e+01 -6.38559837e-01 ... -1.06704488e-01
  -1.66070607e-01 -2.07106818e-01]]


In [ ]:
X=X_new

In [ ]:
print(X.shape)

(4143, 520)


In [ ]:
(no_of_samples, no_of_features)=X.shape

Comments: The original matrix X had 54877 features and ram kept crashing. So, I did data preprocessing to reduce the number of features. As sir suggested, I used PCA with the threshold that 80% of the variance of the original dataset was captured in the processed dataset. The number of features actually comes out to be 514 but still I have taken 520 features.

### Answer b)

In [ ]:
classes, freq = np.unique(y, return_counts=True)

In [ ]:
print("No of classes:", len(classes))

No of classes: 2


In [ ]:
for i in range(len(classes)):
  print("No of samples belonging to class", classes[i], "is", freq[i])

No of samples belonging to class -1.0 is 1933
No of samples belonging to class 1.0 is 2210


There is no class imbalance issue.

### Answer c)

In [ ]:
df=pd.DataFrame(X)

In [ ]:
df

0          1         2         3         4         5         6    \
0     -3.163556   1.354492  0.122609  0.614129 -0.325644 -0.765679 -0.197722   
1      2.944611  -1.853034 -3.423569 -3.943277 -3.041524  3.619496 -3.783854   
2     -2.424726   1.074610 -0.091547  0.004458 -0.401874 -0.567940 -0.354899   
3     -1.785259   0.864062  0.235106  0.025431  0.331659  0.115276 -0.176223   
4     -3.167082   1.337177  0.047686  0.561414 -0.369034 -0.765479 -0.196355   
...         ...        ...       ...       ...       ...       ...       ...   
4138  -1.446324   0.642952 -0.330377 -0.011503 -0.000712 -0.501163  0.366723   
4139  -0.368333   0.016102  0.309881  0.342469 -0.103763 -0.023977  0.059648   
4140   0.889479  -0.901727  2.048592  0.134772 -0.032090  0.193295  0.155847   
4141  -1.755179   0.572071  0.117253  0.273678 -0.116884 -0.404077 -0.023015   
4142  12.216736  14.041053 -0.638560  0.219674  0.336044 -0.271704 -0.013033   

           7         8         9    ...       510       511       512  \
0     0.345623 -0.583530 -0.061195  ...  0.002652  0.023774 -0.007729   
1    -4.853894 -5.449297  0.604696  ...  0.185167  0.087761  0.058403   
2     0.011724 -0.074278  0.295060  ... -0.106116 -0.116623 -0.004392   
3    -0.279108 -0.062008  0.453530  ...  0.017760  0.042083  0.017204   
4     0.298536 -0.632671 -0.046335  ...  0.033637  0.054150  0.051159   
...        ...       ...       ...  ...       ...       ...       ...   
4138  0.138004  0.183987  0.713666  ... -0.419539 -0.059336 -0.254646   
4139 -0.108008  0.518320  0.090268  ...  0.104137 -0.063511  0.099590   
4140 -0.452384  1.554877 -0.544347  ...  0.349073  0.343446 -0.404126   
4141  0.529717 -0.194253  0.237860  ...  0.336364 -0.179108  0.174349   
4142  0.106316 -0.181107 -0.102531  ... -0.094158  0.085157 -0.071463   

           513       514       515       516       517       518       519  
0    -0.011972  0.029302 -0.035175  0.024119  0.017398  0.022221 -0.006849  
1     0.112367 -0.014748  0.195350 -0.123204  0.173635  0.070424  0.027614  
2    -0.148935 -0.113039  0.227284 -0.095412 -0.058572  0.061757  0.256509  
3    -0.010017 -0.027121  0.043902  0.038127  0.005259 -0.063878  0.016864  
4    -0.016397 -0.028229 -0.023845 -0.040030 -0.055646 -0.020364  0.024727  
...        ...       ...       ...       ...       ...       ...       ...  
4138 -0.040410 -0.164844  0.815406  0.671009 -0.376689 -0.560336 -0.078880  
4139 -0.052728  0.021492  0.178274 -0.063507  0.030388 -0.025215  0.080261  
4140  0.802472 -0.964902 -0.072082  0.977557  0.324235 -0.287580 -0.315314  
4141  0.133407  0.140846  0.106979 -0.262924  0.167895  0.053946 -0.155764  
4142  0.295983 -0.004245  0.076728 -0.077094 -0.106704 -0.166071 -0.207107  

[4143 rows x 520 columns]

In [ ]:
df.insert(no_of_features, 'Label', y)

In [ ]:
df

0          1         2         3         4         5         6  \
0     -3.163556   1.354492  0.122609  0.614129 -0.325644 -0.765679 -0.197722   
1      2.944611  -1.853034 -3.423569 -3.943277 -3.041524  3.619496 -3.783854   
2     -2.424726   1.074610 -0.091547  0.004458 -0.401874 -0.567940 -0.354899   
3     -1.785259   0.864062  0.235106  0.025431  0.331659  0.115276 -0.176223   
4     -3.167082   1.337177  0.047686  0.561414 -0.369034 -0.765479 -0.196355   
...         ...        ...       ...       ...       ...       ...       ...   
4138  -1.446324   0.642952 -0.330377 -0.011503 -0.000712 -0.501163  0.366723   
4139  -0.368333   0.016102  0.309881  0.342469 -0.103763 -0.023977  0.059648   
4140   0.889479  -0.901727  2.048592  0.134772 -0.032090  0.193295  0.155847   
4141  -1.755179   0.572071  0.117253  0.273678 -0.116884 -0.404077 -0.023015   
4142  12.216736  14.041053 -0.638560  0.219674  0.336044 -0.271704 -0.013033   

             7         8         9  ...       511       512       513  \
0     0.345623 -0.583530 -0.061195  ...  0.023774 -0.007729 -0.011972   
1    -4.853894 -5.449297  0.604696  ...  0.087761  0.058403  0.112367   
2     0.011724 -0.074278  0.295060  ... -0.116623 -0.004392 -0.148935   
3    -0.279108 -0.062008  0.453530  ...  0.042083  0.017204 -0.010017   
4     0.298536 -0.632671 -0.046335  ...  0.054150  0.051159 -0.016397   
...        ...       ...       ...  ...       ...       ...       ...   
4138  0.138004  0.183987  0.713666  ... -0.059336 -0.254646 -0.040410   
4139 -0.108008  0.518320  0.090268  ... -0.063511  0.099590 -0.052728   
4140 -0.452384  1.554877 -0.544347  ...  0.343446 -0.404126  0.802472   
4141  0.529717 -0.194253  0.237860  ... -0.179108  0.174349  0.133407   
4142  0.106316 -0.181107 -0.102531  ...  0.085157 -0.071463  0.295983   

           514       515       516       517       518       519  Label  
0     0.029302 -0.035175  0.024119  0.017398  0.022221 -0.006849    1.0  
1    -0.014748  0.195350 -0.123204  0.173635  0.070424  0.027614   -1.0  
2    -0.113039  0.227284 -0.095412 -0.058572  0.061757  0.256509   -1.0  
3    -0.027121  0.043902  0.038127  0.005259 -0.063878  0.016864   -1.0  
4    -0.028229 -0.023845 -0.040030 -0.055646 -0.020364  0.024727   -1.0  
...        ...       ...       ...       ...       ...       ...    ...  
4138 -0.164844  0.815406  0.671009 -0.376689 -0.560336 -0.078880   -1.0  
4139  0.021492  0.178274 -0.063507  0.030388 -0.025215  0.080261    1.0  
4140 -0.964902 -0.072082  0.977557  0.324235 -0.287580 -0.315314    1.0  
4141  0.140846  0.106979 -0.262924  0.167895  0.053946 -0.155764   -1.0  
4142 -0.004245  0.076728 -0.077094 -0.106704 -0.166071 -0.207107    1.0  

[4143 rows x 521 columns]

In [ ]:
t1 = df.sample(frac=0.8,random_state=100)
t2 = df.drop(t1.index)

len(t1), len(t2)

(3314, 829)

In [ ]:
t1

0         1         2         3         4         5         6  \
2280 -1.978141  0.589198  0.786112  0.271875 -0.411723  0.117441  0.130675   
2888 -1.871362  0.285709 -0.981533 -0.115179 -0.018796 -0.721802 -0.528681   
1568  0.854090 -1.130812  2.259693  0.490170 -0.079841  0.891291 -0.176710   
3249 -2.063370  0.763066 -1.032347 -0.833967 -0.584380 -0.616016 -0.189024   
3765  0.351286  1.175211  1.670889  0.362812  0.747802 -1.118489 -0.348105   
...        ...       ...       ...       ...       ...       ...       ...   
3994  0.383312  1.162218  1.751956  0.350149  0.755391 -1.152585 -0.355943   
2385  0.362656 -0.895914 -2.102759 -1.578320 -0.215866 -0.326712 -1.613771   
3458 -2.162007  0.650815  0.603762  0.398812 -0.230085 -0.527576  0.056730   
3606 -0.439156 -0.046317  0.821813  0.716538  0.521965  0.344686  0.783902   
1946 -3.168699  1.350222  0.110570  0.605987 -0.338327 -0.776210 -0.185288   

             7         8         9  ...       511       512       513  \
2280 -0.566682 -0.067313  0.297906  ...  0.184632 -0.342961 -0.074727   
2888  0.484127 -0.336386 -0.556121  ...  0.071863 -0.045028  0.071242   
1568 -1.785777  1.024067  0.235464  ...  0.595100 -0.644280  0.018915   
3249  0.020874  0.911860  2.289687  ...  0.046234 -0.141222 -0.390948   
3765  1.289910  1.300334  1.708132  ...  0.044485 -0.089829 -0.187399   
...        ...       ...       ...  ...       ...       ...       ...   
3994  1.244924  1.323388  1.724567  ...  0.011967 -0.126328 -0.202446   
2385  0.281916 -0.202043 -1.172442  ...  0.026638  0.006728 -0.016733   
3458  0.316092 -0.112163 -0.119927  ... -0.123276 -0.019458  0.055739   
3606  0.477981  0.280667  0.504853  ... -0.566485  1.359361 -0.226897   
1946  0.354627 -0.595401 -0.063781  ...  0.054179  0.020682  0.014243   

           514       515       516       517       518       519  Label  
2280 -0.221977 -0.021352  0.002997  0.130308 -0.119302  0.044573   -1.0  
2888 -0.007533 -0.049406  0.024133  0.034523 -0.067504 -0.017471    1.0  
1568  0.647483  0.049491 -0.271779 -0.047100 -0.223357  1.060074    1.0  
3249  0.183533 -0.053263 -0.387359 -0.214758 -0.117201 -0.009616   -1.0  
3765 -0.019341  0.035795  0.158830 -0.022202  0.011162  0.204164   -1.0  
...        ...       ...       ...       ...       ...       ...    ...  
3994 -0.042679  0.067904  0.150570 -0.030854  0.029646  0.143107   -1.0  
2385 -0.060556 -0.042009 -0.010872 -0.007981 -0.044051  0.021935    1.0  
3458  0.019837 -0.235914 -0.005566 -0.392265  0.185622 -0.353924    1.0  
3606 -0.054603  0.536200  0.222981 -0.203072 -0.170949  0.281726    1.0  
1946  0.053786 -0.005625  0.028934  0.043824  0.005872 -0.061578   -1.0  

[3314 rows x 521 columns]

In [ ]:
c_t1, f_t1 = np.unique(t1['Label'], return_counts=True)
c_t2, f_t2 = np.unique(t2['Label'], return_counts=True)

In [ ]:
print("For T1 data:")
for i in range(len(c_t1)):
  print("No of samples belonging to class", c_t1[i], "is", f_t1[i])

print("\nFor T2 data:")
for i in range(len(c_t2)):
  print("No of samples belonging to class", c_t2[i], "is", f_t2[i])

For T1 data:
No of samples belonging to class -1.0 is 1555
No of samples belonging to class 1.0 is 1759

For T2 data:
No of samples belonging to class -1.0 is 378
No of samples belonging to class 1.0 is 451


Both the datasets T1 and T2 have similar class label proportions.

In [ ]:
cols = list(np.arange(0, no_of_features))
X_train = t1[t1.columns[cols]].to_numpy()
X_test = t2[t2.columns[cols]].to_numpy()

In [ ]:
X_train

array([[-1.9781413 ,  0.58919837,  0.78611235, ...,  0.13030788,
        -0.11930248,  0.04457285],
       [-1.8713622 ,  0.28570865, -0.98153328, ...,  0.0345233 ,
        -0.06750414, -0.01747109],
       [ 0.85409033, -1.13081209,  2.25969285, ..., -0.04709963,
        -0.22335692,  1.06007386],
       ...,
       [-2.16200668,  0.65081516,  0.60376202, ..., -0.39226525,
         0.18562227, -0.35392444],
       [-0.43915561, -0.04631684,  0.82181315, ..., -0.20307236,
        -0.17094852,  0.28172559],
       [-3.16869891,  1.35022169,  0.1105697 , ...,  0.04382425,
         0.00587182, -0.06157772]])

In [ ]:
X_test

array([[-2.42472556e+00,  1.07460981e+00, -9.15473283e-02, ...,
        -5.85720272e-02,  6.17565303e-02,  2.56509028e-01],
       [-1.78525920e+00,  8.64062096e-01,  2.35105900e-01, ...,
         5.25877254e-03, -6.38777125e-02,  1.68644332e-02],
       [-3.15131198e+00,  1.34124458e+00,  1.26068021e-01, ...,
        -3.77893550e-02,  1.20958428e-02, -2.41914836e-03],
       ...,
       [ 1.58203966e-01, -3.61287969e-01,  2.63822836e+00, ...,
        -1.12348305e-01,  1.57989282e-03, -2.51199686e-01],
       [-2.34334168e+00,  3.16280251e-01, -1.09120098e+00, ...,
         1.82390843e-02, -1.00974943e-01,  1.49627042e-01],
       [-1.44632430e+00,  6.42951553e-01, -3.30377327e-01, ...,
        -3.76688901e-01, -5.60335817e-01, -7.88800062e-02]])

In [ ]:
response_cols = [no_of_features]
y_train = t1[t1.columns[response_cols]].to_numpy()
y_test = t2[t2.columns[response_cols]].to_numpy()

In [ ]:
y_train

array([[-1.],
       [ 1.],
       [ 1.],
       ...,
       [ 1.],
       [ 1.],
       [-1.]])

In [ ]:
y_test

array([[-1.],
       [-1.],
       [-1.],
       [-1.],
       [ 1.],
       [-1.],
       [ 1.],
       [ 1.],
       [ 1.],
       [-1.],
       [-1.],
       [-1.],
       [-1.],
       [ 1.],
       [ 1.],
       [ 1.],
       [-1.],
       [-1.],
       [-1.],
       [ 1.],
       [-1.],
       [ 1.],
       [ 1.],
       [ 1.],
       [ 1.],
       [ 1.],
       [ 1.],
       [-1.],
       [-1.],
       [-1.],
       [ 1.],
       [-1.],
       [ 1.],
       [ 1.],
       [-1.],
       [-1.],
       [-1.],
       [-1.],
       [ 1.],
       [ 1.],
       [-1.],
       [ 1.],
       [ 1.],
       [-1.],
       [-1.],
       [ 1.],
       [-1.],
       [-1.],
       [ 1.],
       [-1.],
       [-1.],
       [-1.],
       [ 1.],
       [ 1.],
       [ 1.],
       [ 1.],
       [ 1.],
       [ 1.],
       [-1.],
       [-1.],
       [ 1.],
       [-1.],
       [-1.],
       [ 1.],
       [ 1.],
       [ 1.],
       [ 1.],
       [ 1.],
       [-1.],
       [-1.],
       [-1.],
      

### Answer d)

In [ ]:
def metrics_for_evaluation(model, X, y):
    y_pred = model.predict(X)
    confusion_matrix_values = confusion_matrix(y_true=y,y_pred=y_pred)
    accuracy = accuracy_score(y, y_pred)
    precision = precision_score(y, y_pred)
    recall = recall_score(y, y_pred)
    f1 = f1_score(y, y_pred)
    sensitivity = confusion_matrix_values[0,0]/(confusion_matrix_values[0,0]+confusion_matrix_values[0,1])
    specificity = confusion_matrix_values[1,1]/(confusion_matrix_values[1,0]+confusion_matrix_values[1,1])
    return accuracy, precision, recall, f1, sensitivity, specificity

### Part i)

In [ ]:
logistic_l2 = LogisticRegression(penalty='l2', solver='lbfgs', max_iter=1000)
logistic_l2_param_grid = {'C': np.array([0.01, 0.1, 1, 10, 100])}
logistic_l2_cv = GridSearchCV(logistic_l2, logistic_l2_param_grid, cv=5, n_jobs=multiprocessing.cpu_count())
logistic_l2_cv.fit(X_train, y_train)

acc_train, prec_train, rec_train, f_train, sen_train, spec_train=metrics_for_evaluation(logistic_l2_cv, X_train, y_train)
acc_test, prec_test, rec_test, f_test, sen_test, spec_test=metrics_for_evaluation(logistic_l2_cv, X_test, y_test)

print("Logistic regression with L2 regularizer")

print("\nFor train data:")
print("Accuracy =", acc_train)
print("Precision =", prec_train)
print("Recall =", rec_train)
print("F1 Score =", f_train)
print("Sensitivity =", sen_train)
print("Specitivity =", spec_train)

print("\nFor test data:")
print("Accuracy =", acc_test)
print("Precision =", prec_test)
print("Recall =", rec_test)
print("F1 Score =", f_test)
print("Sensitivity =", sen_test)
print("Specitivity =", spec_test)

Logistic regression with L2 regularizer

For train data:
Accuracy = 0.9330114665057333
Precision = 0.9178901576943991
Recall = 0.9596361569073337
F1 Score = 0.9382990550305725
Sensitivity = 0.9028938906752412
Specitivity = 0.9596361569073337

For test data:
Accuracy = 0.8950542822677925
Precision = 0.8888888888888888
Recall = 0.9223946784922394
F1 Score = 0.9053318824809575
Sensitivity = 0.8624338624338624
Specitivity = 0.9223946784922394


In [ ]:
svm_l2 = SVC(kernel='linear')
svm_l2_param_grid = {'C':np.array([0.01, 0.1, 1, 10, 100])}
svm_l2_cv = GridSearchCV(svm_l2, svm_l2_param_grid, cv=5,n_jobs=multiprocessing.cpu_count())
svm_l2_cv.fit(X_train,y_train)

acc_train, prec_train, rec_train, f_train, sen_train, spec_train=metrics_for_evaluation(svm_l2_cv, X_train, y_train)
acc_test, prec_test, rec_test, f_test, sen_test, spec_test=metrics_for_evaluation(svm_l2_cv, X_test, y_test)

print("Soft-margin SVM with L2 regularizer")

print("\nFor train data:")
print("Accuracy =", acc_train)
print("Precision =", prec_train)
print("Recall =", rec_train)
print("F1 Score =", f_train)
print("Sensitivity =", sen_train)
print("Specitivity =", spec_train)

print("\nFor test data:")
print("Accuracy =", acc_test)
print("Precision =", prec_test)
print("Recall =", rec_test)
print("F1 Score =", f_test)
print("Sensitivity =", sen_test)
print("Specitivity =", spec_test)

Soft-margin SVM with L2 regularizer

For train data:
Accuracy = 0.9417622208811104
Precision = 0.926006528835691
Recall = 0.9675952245594087
F1 Score = 0.9463441757019738
Sensitivity = 0.912540192926045
Specitivity = 0.9675952245594087

For test data:
Accuracy = 0.8986731001206273
Precision = 0.8997821350762527
Recall = 0.9157427937915743
F1 Score = 0.9076923076923077
Sensitivity = 0.8783068783068783
Specitivity = 0.9157427937915743


In [ ]:
svm_l1 = LinearSVC(penalty='l1', dual=False, class_weight = 'balanced')
svm_l1_param_grid = {'C': np.array([0.01, 0.1, 1, 10, 100])}
svm_l1_cv = GridSearchCV(svm_l1, svm_l1_param_grid, cv=5,n_jobs=multiprocessing.cpu_count())
svm_l1_cv.fit(X_train,y_train)

acc_train, prec_train, rec_train, f_train, sen_train, spec_train=metrics_for_evaluation(svm_l1_cv, X_train, y_train)
acc_test, prec_test, rec_test, f_test, sen_test, spec_test=metrics_for_evaluation(svm_l1_cv, X_test, y_test)

print("Soft-margin SVM with L1 regularizer")

print("\nFor train data:")
print("Accuracy =", acc_train)
print("Precision =", prec_train)
print("Recall =", rec_train)
print("F1 Score =", f_train)
print("Sensitivity =", sen_train)
print("Specitivity =", spec_train)

print("\nFor test data:")
print("Accuracy =", acc_test)
print("Precision =", prec_test)
print("Recall =", rec_test)
print("F1 Score =", f_test)
print("Sensitivity =", sen_test)
print("Specitivity =", spec_test)

Soft-margin SVM with L1 regularizer

For train data:
Accuracy = 0.9269764634882317
Precision = 0.9263631253513209
Recall = 0.9368959636156907
F1 Score = 0.93159977388355
Sensitivity = 0.9157556270096463
Specitivity = 0.9368959636156907

For test data:
Accuracy = 0.887816646562123
Precision = 0.9049773755656109
Recall = 0.8869179600886918
F1 Score = 0.8958566629339305
Sensitivity = 0.8888888888888888
Specitivity = 0.8869179600886918


In [ ]:
svm_rbf = SVC(kernel='rbf')
svm_rbf_param_grid = {'gamma': np.array([ 0.01, 0.1, 1, 10, 100])}
svm_rbf_cv = GridSearchCV(svm_rbf, svm_rbf_param_grid, cv=5,n_jobs=multiprocessing.cpu_count())
svm_rbf_cv.fit(X_train,y_train)

acc_train, prec_train, rec_train, f_train, sen_train, spec_train=metrics_for_evaluation(svm_rbf_cv, X_train, y_train)
acc_test, prec_test, rec_test, f_test, sen_test, spec_test=metrics_for_evaluation(svm_rbf_cv, X_test, y_test)

print("Kernel SVM with RBF kernel")

print("\nFor train data:")
print("Accuracy =", acc_train)
print("Precision =", prec_train)
print("Recall =", rec_train)
print("F1 Score =", f_train)
print("Sensitivity =", sen_train)
print("Specitivity =", spec_train)

print("\nFor test data:")
print("Accuracy =", acc_test)
print("Precision =", prec_test)
print("Recall =", rec_test)
print("F1 Score =", f_test)
print("Sensitivity =", sen_test)
print("Specitivity =", spec_test)

Kernel SVM with RBF kernel

For train data:
Accuracy = 0.9378394689197345
Precision = 0.9136920618007459
Recall = 0.9749857873791927
F1 Score = 0.9433443344334433
Sensitivity = 0.8958199356913183
Specitivity = 0.9749857873791927

For test data:
Accuracy = 0.8854041013268998
Precision = 0.8517786561264822
Recall = 0.9556541019955654
F1 Score = 0.9007314524555903
Sensitivity = 0.8015873015873016
Specitivity = 0.9556541019955654


In [ ]:
knn = KNeighborsClassifier()
k_range = list(range(1, 31))
knn_cv = GridSearchCV(knn, {'n_neighbors': k_range}, cv=5, scoring='accuracy', n_jobs=-1)
knn_cv.fit(X_train, y_train)

acc_train, prec_train, rec_train, f_train, sen_train, spec_train=metrics_for_evaluation(knn_cv, X_train, y_train)
acc_test, prec_test, rec_test, f_test, sen_test, spec_test=metrics_for_evaluation(knn_cv, X_test, y_test)

print("KNN")

print("\nFor train data:")
print("Accuracy =", acc_train)
print("Precision =", prec_train)
print("Recall =", rec_train)
print("F1 Score =", f_train)
print("Sensitivity =", sen_train)
print("Specitivity =", spec_train)

print("\nFor test data:")
print("Accuracy =", acc_test)
print("Precision =", prec_test)
print("Recall =", rec_test)
print("F1 Score =", f_test)
print("Sensitivity =", sen_test)
print("Specitivity =", spec_test)

KNN

For train data:
Accuracy = 0.9990947495473748
Precision = 0.9988636363636364
Recall = 0.9994314951677089
F1 Score = 0.9991474850809889
Sensitivity = 0.9987138263665595
Specitivity = 0.9994314951677089

For test data:
Accuracy = 0.8829915560916767
Precision = 0.9022727272727272
Recall = 0.8802660753880266
F1 Score = 0.8911335578002244
Sensitivity = 0.8862433862433863
Specitivity = 0.8802660753880266


In [ ]:
dt = DecisionTreeClassifier()
dt_params = {'min_weight_fraction_leaf': np.linspace(0.0, 0.5, 11)}
dt_cv = GridSearchCV(dt, dt_params, cv=5, scoring='accuracy', n_jobs=-1)
dt_cv.fit(X_train, y_train)

acc_train, prec_train, rec_train, f_train, sen_train, spec_train=metrics_for_evaluation(dt_cv, X_train, y_train)
acc_test, prec_test, rec_test, f_test, sen_test, spec_test=metrics_for_evaluation(dt_cv, X_test, y_test)

print("Decision Tree")

print("\nFor train data:")
print("Accuracy =", acc_train)
print("Precision =", prec_train)
print("Recall =", rec_train)
print("F1 Score =", f_train)
print("Sensitivity =", sen_train)
print("Specitivity =", spec_train)

print("\nFor test data:")
print("Accuracy =", acc_test)
print("Precision =", prec_test)
print("Recall =", rec_test)
print("F1 Score =", f_test)
print("Sensitivity =", sen_test)
print("Specitivity =", spec_test)

Decision Tree

For train data:
Accuracy = 0.9990947495473748
Precision = 0.9988636363636364
Recall = 0.9994314951677089
F1 Score = 0.9991474850809889
Sensitivity = 0.9987138263665595
Specitivity = 0.9994314951677089

For test data:
Accuracy = 0.8311218335343787
Precision = 0.859122401847575
Recall = 0.8248337028824834
F1 Score = 0.841628959276018
Sensitivity = 0.8386243386243386
Specitivity = 0.8248337028824834


In [ ]:
rf = RandomForestClassifier()
rf_params = {'n_estimators': list(range(50, 251, 50))}
rf_cv = GridSearchCV(rf, rf_params, cv=5, scoring='accuracy', n_jobs=-1)
rf_cv.fit(X_train, y_train)

acc_train, prec_train, rec_train, f_train, sen_train, spec_train=metrics_for_evaluation(rf_cv, X_train, y_train)
acc_test, prec_test, rec_test, f_test, sen_test, spec_test=metrics_for_evaluation(rf_cv, X_test, y_test)

print("Random Forest")

print("\nFor train data:")
print("Accuracy =", acc_train)
print("Precision =", prec_train)
print("Recall =", rec_train)
print("F1 Score =", f_train)
print("Sensitivity =", sen_train)
print("Specitivity =", spec_train)

print("\nFor test data:")
print("Accuracy =", acc_test)
print("Precision =", prec_test)
print("Recall =", rec_test)
print("F1 Score =", f_test)
print("Sensitivity =", sen_test)
print("Specitivity =", spec_test)

Random Forest

For train data:
Accuracy = 0.9990947495473748
Precision = 0.9988636363636364
Recall = 0.9994314951677089
F1 Score = 0.9991474850809889
Sensitivity = 0.9987138263665595
Specitivity = 0.9994314951677089

For test data:
Accuracy = 0.8866103739445115
Precision = 0.8606060606060606
Recall = 0.9445676274944568
F1 Score = 0.9006342494714589
Sensitivity = 0.8174603174603174
Specitivity = 0.9445676274944568


In [ ]:
logistic_l1 = LogisticRegression(penalty='l1', solver='liblinear', max_iter=1000)
logistic_l1_param_grid = {'C': np.array([0.01, 0.1, 1, 10, 100])}
logistic_l1_cv = GridSearchCV(logistic_l1, logistic_l1_param_grid, cv=5,n_jobs=multiprocessing.cpu_count())
logistic_l1_cv.fit(X_train,y_train)

acc_train, prec_train, rec_train, f_train, sen_train, spec_train=metrics_for_evaluation(logistic_l1_cv, X_train, y_train)
acc_test, prec_test, rec_test, f_test, sen_test, spec_test=metrics_for_evaluation(logistic_l1_cv, X_test, y_test)

print("Logistic regression with L1 regularizer")

print("\nFor train data:")
print("Accuracy =", acc_train)
print("Precision =", prec_train)
print("Recall =", rec_train)
print("F1 Score =", f_train)
print("Sensitivity =", sen_train)
print("Specitivity =", spec_train)

print("\nFor test data:")
print("Accuracy =", acc_test)
print("Precision =", prec_test)
print("Recall =", rec_test)
print("F1 Score =", f_test)
print("Sensitivity =", sen_test)
print("Specitivity =", spec_test)

Logistic regression with L1 regularizer

For train data:
Accuracy = 0.9450814725407363
Precision = 0.9349145063430778
Recall = 0.9636156907333713
F1 Score = 0.9490481522956328
Sensitivity = 0.9241157556270096
Specitivity = 0.9636156907333713

For test data:
Accuracy = 0.8890229191797346
Precision = 0.8980044345898004
Recall = 0.8980044345898004
F1 Score = 0.8980044345898004
Sensitivity = 0.8783068783068783
Specitivity = 0.8980044345898004


In [ ]:
modelName = ['Logistic Regression (L2 Regularizer)','Logistic Regression (L1 Regularizer)','Soft-margin SVM (L2 Regularizer)','Soft-margin SVM (L1 Regularizer)',
'Kernel SVM (RBF kernel)','KNN','Decision Tree','Random Forest']
i = 0
dataframe_best_params = pd.DataFrame({'Model Name':[],'Best parameters':[],'Best score':[]})
for model in [logistic_l2_cv, logistic_l1_cv, svm_l2_cv, svm_l1_cv, svm_rbf_cv, knn_cv, dt_cv, rf_cv]:
  dataframe_best_params.loc[len(dataframe_best_params.index)] = [modelName[i],model.best_params_,model.best_score_]
  i +=1

In [ ]:
print("Table for best parameters obtained after hyperparameter tuning in each model")
dataframe_best_params

Table for best parameters obtained after hyperparameter tuning in each model


Model Name                    Best parameters  \
0  Logistic Regression (L2 Regularizer)                         {'C': 0.1}   
1  Logistic Regression (L1 Regularizer)                         {'C': 1.0}   
2      Soft-margin SVM (L2 Regularizer)                         {'C': 0.1}   
3      Soft-margin SVM (L1 Regularizer)                         {'C': 0.1}   
4               Kernel SVM (RBF kernel)                    {'gamma': 0.01}   
5                                   KNN                 {'n_neighbors': 1}   
6                         Decision Tree  {'min_weight_fraction_leaf': 0.0}   
7                         Random Forest              {'n_estimators': 250}   

   Best score  
0    0.884132  
1    0.873874  
2    0.880507  
3    0.872364  
4    0.882319  
5    0.854860  
6    0.806885  
7    0.871458

Observations: We can observe the best parameters for each model after hyperparameter tuning. Scores lie between 0 and 1. The higher the score, the better the model has performed while training with respect to that parameter.  

### Answer e)

In [ ]:
dataframe_train = pd.DataFrame({'Model Name':[],'Accuracy':[],'Precision':[],'Recall':[],'F1 Score':[],'Specificity':[],'Senstivity':[]})
dataframe_test = pd.DataFrame({'Model Name':[],'Accuracy':[],'Precision':[],'Recall':[],'F1 Score':[],'Specificity':[],'Senstivity':[]})
modelName = ['Logistic Regression (L2 Regularizer)','Logistic Regression (L1 Regularizer)','Soft-margin SVM (L2 Regularizer)','Soft-margin SVM (L1 Regularizer)',
'Kernel SVM (RBF kernel)','KNN','Decision Tree','Random Forest']
i = 0
for model in [logistic_l2_cv, logistic_l1_cv, svm_l2_cv, svm_l1_cv, svm_rbf_cv, knn_cv, dt_cv, rf_cv]:
  accuracy_train, precision_train, recall_train, f1_train, sensitivity_train, specificity_train = metrics_for_evaluation(model, X_train, y_train)
  accuracy_test, precision_test, recall_test, f1_test, sensitivity_test, specificity_test = metrics_for_evaluation(model, X_test, y_test)
  dataframe_train.loc[len(dataframe_train.index)] = [modelName[i], accuracy_train, precision_train, recall_train, f1_train, sensitivity_train, specificity_train]
  dataframe_test.loc[len(dataframe_test.index)] = [modelName[i], accuracy_test, precision_test, recall_test, f1_test, sensitivity_test, specificity_test]
  i += 1

In [ ]:
print("Tabulated accuracy, precision, recall, specificity and senstivity values for training set")
dataframe_train

Tabulated accuracy, precision, recall, specificity and senstivity values for training set


Model Name  Accuracy  Precision    Recall  \
0  Logistic Regression (L2 Regularizer)  0.933011   0.917890  0.959636   
1  Logistic Regression (L1 Regularizer)  0.945081   0.934915  0.963616   
2      Soft-margin SVM (L2 Regularizer)  0.941762   0.926007  0.967595   
3      Soft-margin SVM (L1 Regularizer)  0.926976   0.926363  0.936896   
4               Kernel SVM (RBF kernel)  0.937839   0.913692  0.974986   
5                                   KNN  0.999095   0.998864  0.999431   
6                         Decision Tree  0.999095   0.998864  0.999431   
7                         Random Forest  0.999095   0.998864  0.999431   

   F1 Score  Specificity  Senstivity  
0  0.938299     0.902894    0.959636  
1  0.949048     0.924116    0.963616  
2  0.946344     0.912540    0.967595  
3  0.931600     0.915756    0.936896  
4  0.943344     0.895820    0.974986  
5  0.999147     0.998714    0.999431  
6  0.999147     0.998714    0.999431  
7  0.999147     0.998714    0.999431

In [ ]:
print("Tabulated accuracy, precision, recall, specificity and senstivity values for test set")
dataframe_test

Tabulated accuracy, precision, recall, specificity and senstivity values for test set


Model Name  Accuracy  Precision    Recall  \
0  Logistic Regression (L2 Regularizer)  0.895054   0.888889  0.922395   
1  Logistic Regression (L1 Regularizer)  0.889023   0.898004  0.898004   
2      Soft-margin SVM (L2 Regularizer)  0.898673   0.899782  0.915743   
3      Soft-margin SVM (L1 Regularizer)  0.887817   0.904977  0.886918   
4               Kernel SVM (RBF kernel)  0.885404   0.851779  0.955654   
5                                   KNN  0.882992   0.902273  0.880266   
6                         Decision Tree  0.831122   0.859122  0.824834   
7                         Random Forest  0.886610   0.860606  0.944568   

   F1 Score  Specificity  Senstivity  
0  0.905332     0.862434    0.922395  
1  0.898004     0.878307    0.898004  
2  0.907692     0.878307    0.915743  
3  0.895857     0.888889    0.886918  
4  0.900731     0.801587    0.955654  
5  0.891134     0.886243    0.880266  
6  0.841629     0.838624    0.824834  
7  0.900634     0.817460    0.944568

### Answer f)

Both L1 and L2 regularizers are used to avoid overfitting in linear models such as logistic regression and SVM. However, L1 regularization has a tendency to result in sparse models because it forces some of the coefficients to be exactly zero. On the other hand, L2 regularization does not force coefficients to be exactly zero but rather shrinks them towards zero.

L1 regularization is often better suited for situations where there are many irrelevant features, as it allows for feature selection by setting some coefficients to zero. L2 regularization, on the other hand, is better suited for situations where all the features are relevant, as it avoids overfitting by shrinking all the coefficients towards zero.

In practice, whether to use L1 or L2 regularization depends on the specific problem and data at hand. If the goal is to perform feature selection and obtain a sparse model, then L1 regularization should be used. If all the features are expected to be relevant, then L2 regularization should be used. In some cases, it may be useful to try both regularization techniques and compare their performance on the validation set.

Overall, L1 regularization is a useful tool for feature selection in high-dimensional datasets where only a few features are expected to be relevant. However, it is important to note that L1 regularization may also result in a loss of information if important features are mistakenly set to zero.

Performance of the models obtained using L1 and L2 regularizers can be compared from the table in part e. We also have to choose the evaluation metrics from accuracy, precision, recall, f1 score, sensitivity and specificity and decide which are important to us based on the problem statement related to the dataset. After deciding the metrics, we have to check which model gives higher values (close to 1) of those metrics.

For example, if our chosen evaluation metric is accuracy then we can see that Logistic Regression and soft-margin SVM with L2 regularizer performs better than with L1 regularizer. If our chosen metric is precision then L2 regularizer gives better result than L1 regularizer. This is based on observations of test data.